lib/mongo/server/connection.rb

```diff
@@ -155,7 +155,15 @@ def ping
 
       def deliver(messages)
         write(messages)
-        messages.last.replyable? ? read : nil
+
+        if messages.last.replyable?
+          # Protection against returning the response to a previous request. See
+          # RUBY-1117
+          reply = read until reply && reply.response_to == messages.last.request_id
```

Comment:
@estolfo Ok I'm not sure what you mean in your first paragraph, why would the `socket_timeout` and `max_time_ms` options help here ? To reduce the change the query leaks on the main thread ?

About the `disconnect!` indeed we missed this one, so actually if we put the raise in `deserialize`, it should work and recover properly without our change to connection_pool. We just tried and it works as expected, it raises in case of error and the socket recovers at the next query.

And so with this fix, we would have to disconnect every mongo socket each time we kill a thread (or EM.stop) if we want to be safe, right ?

This should still be improved though IMO, I'm trying to find if there is a way, basically if the read is interrupted, we want to close the socket, the problem is that killing a thread calls every `ensure` block (effectively checking in the connection back in the pool to be reused) BUT it doesn't call any `rescue Exception` block, which prevents the socket from being closed. If we can find a way to close the socket also when the thread is killed and not only on Exception, that would solve this problem entirely. I had a look at `handle_interrupt` and it seems to be handy to protect your code against exception (you can defer them to interrupt your thread at a later point, that's why concurrent_ruby uses it for it's thread pool: https://github.com/mperham/connection_pool/blob/master/lib/connection_pool.rb#L59) but it doesn't seem to do anything with `kill`, as opposed to what they say in the Ruby doc.